File: sdks/local-rp/dart/test/conformance/callback_box_test.dart
import 'package:linkkeys_local_rp/linkkeys_local_rp.dart';
import 'package:linkkeys_local_rp/src/wire/codec.dart';
import 'package:test/test.dart';

import '../testutil/fixtures.dart';

List<AeadSuite> _parseAllowedSuites(Map<String, dynamic> c) {
  final out = <AeadSuite>[];
  for (final s in (c['allowed_suites'] as List).cast<String>()) {
    final suite = AeadSuite.parse(s);
    if (suite == null) {
      throw StateError('unregistered suite id in fixture: $s');
    }
    out.add(suite);
  }
  return out;
}

void main() {
  group('callback_box.json conformance', () {
    test('positive cases open via SDK dependency', () async {
      final d = loadJson('callback_box.json') as Map<String, dynamic>;
      final cases = (d['positive_cases'] as List).cast<Map<String, dynamic>>();
      expect(cases.length, equals(2));

      for (final c in cases) {
        final headerBytes = hex(c['header_cbor_hex'] as String);
        final ciphertext = hex(c['ciphertext_hex'] as String);
        final decryptKey = hex(c['decrypt_private_key_hex'] as String);
        final allowed = _parseAllowedSuites(c);

        final encrypted = LocalRpEncryptedCallback(
            header: headerBytes, ciphertext: ciphertext);
        final opened =
            await LocalRp.openLocalRpCallback(encrypted, decryptKey, allowed);

        expect(opened.header.suite, equals(c['suite']));
        expect(opened.header.fingerprint, equals(c['fingerprint']));
        expect(opened.header.nonce, equals(hex(c['nonce_hex'] as String)));
        expect(opened.header.state, equals(hex(c['state_hex'] as String)));
        expect(opened.header.issuedAt, equals(c['issued_at']));
        expect(opened.header.expiresAt, equals(c['expires_at']));

        final plaintext =
            Codec.encodeSignedLocalRpCallbackPayload(opened.signedPayload);
        expect(plaintext, equals(hex(c['plaintext_cbor_hex'] as String)));
      }
    });

    test('negative cases fail', () async {
      final d = loadJson('callback_box.json') as Map<String, dynamic>;
      final cases = (d['negative_cases'] as List).cast<Map<String, dynamic>>();
      expect(cases.length, equals(13));

      for (final c in cases) {
        final headerBytes = hex(c['header_cbor_hex'] as String);
        final ciphertext = hex(c['ciphertext_hex'] as String);
        final decryptKey = hex(c['decrypt_private_key_hex'] as String);
        final allowed = _parseAllowedSuites(c);

        final encrypted = LocalRpEncryptedCallback(
            header: headerBytes, ciphertext: ciphertext);
        await expectLater(
          LocalRp.openLocalRpCallback(encrypted, decryptKey, allowed),
          throwsA(anything),
          reason: 'negative case unexpectedly opened: ${c['name']}',
        );
      }
    });
  });
}

File: sdks/local-rp/php/tests/conformance/CallbackBoxTest.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../bootstrap.php';

use Csilgen\Generated\LocalRpEncryptedCallback;
use LinkKeys\LocalRp\Crypto;
use LinkKeys\LocalRp\LocalRp;
use LinkKeys\LocalRp\LocalRpError;
use LinkKeys\LocalRp\Wire;

$vectors = loadJson(__DIR__ . '/../../../conformance/callback_box.json');

foreach ($vectors['positive_cases'] as $case) {
    TestKit::test('callback_box.positive.' . $case['suite'], function () use ($case) {
        $encrypted = new LocalRpEncryptedCallback([
            'header' => hexToBytes($case['header_cbor_hex']),
            'ciphertext' => hexToBytes($case['ciphertext_hex']),
        ]);
        $decryptKey = hexToBytes($case['decrypt_private_key_hex']);
        $allowedSuites = $case['allowed_suites'];

        [$header, $signedPayload] = LocalRp::openLocalRpCallback($encrypted, $decryptKey, $allowedSuites);

        TestKit::assertEquals($case['suite'], $header->suite);
        TestKit::assertEquals($case['fingerprint'], $header->fingerprint);
        TestKit::assertEquals(hexToBytes($case['nonce_hex']), $header->nonce);
        TestKit::assertEquals(hexToBytes($case['state_hex']), $header->state);

        $expectedSignedPayload = Wire::decodeSignedLocalRpCallbackPayload(hexToBytes($case['plaintext_cbor_hex']));
        TestKit::assertEquals($expectedSignedPayload->payload, $signedPayload->payload);
        TestKit::assertEquals($expectedSignedPayload->signingKeyId, $signedPayload->signingKeyId);
        TestKit::assertEquals($expectedSignedPayload->signature, $signedPayload->signature);

        // Independent HKDF/AAD-context check (README: "publish so you can
        // unit-test your own HKDF derivation independent of a full decrypt").
        [$aeadKey, $kdfContext] = LocalRp::localRpCallbackKdf(
            $case['suite'],
            hexToBytes($case['ephemeral_public_key_hex']),
            hexToBytes($case['recipient_public_key_hex']),
            Crypto::x25519DiffieHellman($decryptKey, hexToBytes($case['ephemeral_public_key_hex']))
        );
        TestKit::assertEquals(hexToBytes($case['kdf_context_hex']), $kdfContext, 'kdf_context_hex mismatch');
        $expectedAad = hexToBytes($case['kdf_context_hex']) . hexToBytes($case['header_cbor_hex']);
        TestKit::assertEquals(hexToBytes($case['aad_hex']), $expectedAad, 'aad_hex mismatch');
    });
}

foreach ($vectors['negative_cases'] as $case) {
    TestKit::test('callback_box.negative.' . $case['name'], function () use ($case) {
        $encrypted = new LocalRpEncryptedCallback([
            'header' => hexToBytes($case['header_cbor_hex']),
            'ciphertext' => hexToBytes($case['ciphertext_hex']),
        ]);
        $decryptKey = hexToBytes($case['decrypt_private_key_hex']);
        $allowedSuites = $case['allowed_suites'];

        TestKit::assertThrows(function () use ($encrypted, $decryptKey, $allowedSuites) {
            LocalRp::openLocalRpCallback($encrypted, $decryptKey, $allowedSuites);
        }, $case['name'] . ' should fail: ' . ($case['description'] ?? ''));
    });
}

exit(TestKit::summary('CallbackBoxTest') === 0 ? 0 : 1);
